core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,295 @@
+      currentTTSTask = null
+    } else if (tts.isSpeaking) {
+      if (tts.stop() == SUCCESS) {
+        tts.setOnUtteranceProgressListener(null)
+        onSpeakingListener.onSpeakingEnded()
+      }
+    } else {
+      val locale = iSO3ToLocale(zimReaderContainer.language)
+      if ("mul" == zimReaderContainer.language) {
+        Log.d(
+          TAG_KIWIX, "TextToSpeech: disabled " +
+            zimReaderContainer.language
+        )
+        context.toast(R.string.tts_not_enabled, Toast.LENGTH_LONG)
+        return
+      }
+      if (locale == null || isMissingOrUnsupportedLanguage(tts.isLanguageAvailable(locale))) {
+        Log.d(
+          TAG_KIWIX, "TextToSpeech: language not supported:  ${zimReaderContainer.language}"
+        )
+        context.toast(R.string.tts_lang_not_supported, Toast.LENGTH_LONG)
+      } else {
+        tts.language = locale
+        if (getFeatures(tts).contains(Engine.KEY_FEATURE_NOT_INSTALLED)) {
+          context.toast(R.string.tts_lang_not_supported, Toast.LENGTH_LONG)
+          return
+        }
+        if (requestAudioFocus()) {
+          loadURL(webView)
+        }
+      }
+    }
+  }
+
+  private fun isMissingOrUnsupportedLanguage(languageAvailabilityResult: Int): Boolean =
+    languageAvailabilityResult == LANG_MISSING_DATA ||
+      languageAvailabilityResult == LANG_NOT_SUPPORTED
+
+  private fun getFeatures(
+    tts: TextToSpeech?
+  ): Set<String> = tts!!.voice.features
+
+  private fun loadURL(webView: WebView) {
+    // We use JavaScript to get the content of the page conveniently, earlier making some
+    // changes in the page
+    webView.loadUrl(
+      """
+        javascript:body = document.getElementsByTagName('body')[0].cloneNode(true);
+        // Remove some elements that are shouldn't be read (table of contents,
+        // references numbers, thumbnail captions, duplicated title, etc.)
+        "toRemove = body.querySelectorAll('sup.reference, #toc, .thumbcaption, 
+        "    title, .navbox');
+        "Array.prototype.forEach.call(toRemove, function(elem) {
+        "    elem.parentElement.removeChild(elem);
+        "});
+        "tts.speakAloud(body.innerText);
+        """
+    )
+  }
+
+  fun stop() {
+    tts.let {
+      {
+        if (it.stop() == SUCCESS) {
+          currentTTSTask = null
+          it.setOnUtteranceProgressListener(null)
+          onSpeakingListener.onSpeakingEnded()
+        }
+      }
+    }
+  }
+
+  private fun requestAudioFocus(): Boolean {
+    val audioFocusRequest = am.requestAudioFocus(
+      onAudioFocusChangeListener, AudioManager.STREAM_MUSIC,
+      AudioManager.AUDIOFOCUS_GAIN
+    )
+    Log.d(TAG_KIWIX, "Audio Focus Requested")
+    synchronized(focusLock) {
+      return@requestAudioFocus audioFocusRequest ==
+        AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  fun pauseOrResume() {
+    currentTTSTask?.let {
+      if (it.paused) {
+        if (!requestAudioFocus()) return@pauseOrResume
```

Comment:
this is strange, why return? execution stops here if it is paused whether or not an audio focus request was successful

Reply:
```
 if (it.paused) {
      return@pauseOrResume
    }
 it.pause()
```
Is this right?

Reply:
```
if (it.paused){
requestAudioFocus()
}
else{
it.pause()
}
```